File: sdk/flutter/aio_tracker.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;

/// All-in-One Analytics SDK cho Flutter
/// Sử dụng:
/// ```dart
/// void main() async {
///   WidgetsFlutterBinding.ensureInitialized();
///   await AIOTracker.init(
///     serverUrl: 'http://192.168.1.100:3000',
///     projectId: 'my-flutter-shop',
///     appName: 'My Shop Flutter App',
///   );
///   runApp(MyApp());
/// }
/// ```
class AIOTracker {
  static late String _serverUrl;
  static late String _projectId;
  static late String _appName;
  static String? _sessionId;
  static String? _userId;
  static Timer? _heartbeatTimer;

  static bool _initialized = false;

  /// Khởi tạo Tracker cho Flutter App
  static Future<void> init({
    required String serverUrl,
    required String projectId,
    String? appName,
    String? initialUserId,
  }) async {
    _serverUrl = serverUrl.replaceAll(RegExp(r'/+$'), '');
    _projectId = projectId;
    _appName = appName ?? projectId;
    _userId = initialUserId;
    _sessionId = 'fl_${DateTime.now().millisecondsSinceEpoch}_${(1000 + (DateTime.now().microsecond % 9000))}';

    _initialized = true;

    // Gửi sự kiện mở app đầu tiên
    await trackScreen('AppLaunch');

    // Bắt đầu timer heartbeat mỗi 30s
    _heartbeatTimer?.cancel();
    _heartbeatTimer = Timer.periodic(const Duration(seconds: 30), (timer) {
      _sendHeartbeat();
    });
  }

  /// Theo dõi Màn hình (Screen View)
  static Future<void> trackScreen(String screenName) async {
    if (!_initialized) return;

    final osName = Platform.isAndroid ? 'Android' : (Platform.isIOS ? 'iOS' : Platform.operatingSystem);
    final platform = Platform.isIOS ? 'mobile_ios' : 'mobile_android';

    await _send('/api/v1/track', {
      'projectId': _projectId,
      'projectName': _appName,
      'sessionId': _sessionId,
      'userId': _userId,
      'platform': platform,
      'eventType': 'screen_view',
      'eventName': 'Screen: $screenName',
      'screen': screenName,
      'os': osName,
      'osVersion': Platform.operatingSystemVersion,
      'browser': 'Flutter Engine',
      'clientSdk': 'flutter',
    });
  }

  /// Theo dõi Sự kiện tùy chọn (Nút bấm, Mua hàng, v.v.)
  static Future<void> trackEvent(String eventName, {Map<String, dynamic>? properties}) async {
    if (!_initialized) return;

    final platform = Platform.isIOS ? 'mobile_ios' : 'mobile_android';

    await _send('/api/v1/track', {
      'projectId': _projectId,
      'sessionId': _sessionId,
      'userId': _userId,
      'platform': platform,
      'eventType': 'custom',
      'eventName': eventName,
      'properties': properties ?? {},
      'clientSdk': 'flutter',
    });
  }

  /// Gửi báo cáo lỗi Crash / Exception
  static Future<void> trackError(dynamic error, dynamic stackTrace, {String? context}) async {
    if (!_initialized) return;

    final platform = Platform.isIOS ? 'mobile_ios' : 'mobile_android';

    await _send('/api/v1/track', {
      'projectId': _projectId,
      'sessionId': _sessionId,
      'platform': platform,
      'eventType': 'error',
      'eventName': 'Flutter Exception',
      'path': context ?? 'Unhandled',
      'error': {
        'message': error.toString(),
        'stack': stackTrace.toString(),
      },
      'clientSdk': 'flutter',
    });
  }

  /// Thiết lập ID người dùng (sau khi đăng nhập)
  static void identify(String userId) {
    _userId = userId;
  }

  static Future<void> _sendHeartbeat() async {
    if (!_initialized) return;
    await _send('/api/v1/heartbeat', {
      'projectId': _projectId,
      'sessionId': _sessionId,
    });
  }

  static Future<void> _send(String endpoint, Map<String, dynamic> data) async {
    try {
      final uri = Uri.parse('$_serverUrl$endpoint');
      await http.post(
        uri,
        headers: {
          'Content-Type': 'application/json',
          'X-Platform': Platform.isIOS ? 'mobile_ios' : 'mobile_android',
          'X-Client-Sdk': 'flutter',
        },
        body: jsonEncode(data),
      ).timeout(const Duration(seconds: 4));
    } catch (_) {
      // Âm thầm bỏ qua lỗi kết nối mạng để không ảnh hưởng đến app
    }
  }

  static void dispose() {
    _heartbeatTimer?.cancel();
  }
}

File: sdk/react-native/aioTracker.js
import { Platform, Dimensions } from 'react-native';

/**
 * All-in-One Analytics SDK cho React Native
 * Cách dùng:
 * ```js
 * import AIOTracker from './aioTracker';
 * AIOTracker.init({
 *   serverUrl: 'http://192.168.1.100:3000',
 *   projectId: 'my-rn-app',
 *   appName: 'My React Native App'
 * });
 * ```
 */
class AIOTrackerClient {
  constructor() {
    this.serverUrl = '';
    this.projectId = '';
    this.appName = '';
    this.sessionId = '';
    this.userId = null;
    this.timer = null;
    this.initialized = false;
  }

  init({ serverUrl, projectId, appName, userId }) {
    this.serverUrl = serverUrl.replace(/\/+$/, '');
    this.projectId = projectId;
    this.appName = appName || projectId;
    this.userId = userId || null;
    this.sessionId = 'rn_' + Math.random().toString(36).substring(2, 10) + Date.now().toString(36);
    this.initialized = true;

    // Theo dõi màn hình khởi động
    this.trackScreen('AppLaunch');

    // Heartbeat
    if (this.timer) clearInterval(this.timer);
    this.timer = setInterval(() => this.heartbeat(), 30000);
  }

  trackScreen(screenName) {
    if (!this.initialized) return;
    const { width, height } = Dimensions.get('window');
    const platform = Platform.OS === 'ios' ? 'mobile_ios' : 'mobile_android';

    this.send('/api/v1/track', {
      projectId: this.projectId,
      projectName: this.appName,
      sessionId: this.sessionId,
      userId: this.userId,
      platform: platform,
      eventType: 'screen_view',
      eventName: `Screen: ${screenName}`,
      screen: screenName,
      os: Platform.OS === 'ios' ? 'iOS' : 'Android',
      osVersion: String(Platform.Version),
      browser: 'React Native',
      screenRes: `${Math.round(width)}x${Math.round(height)}`,
      clientSdk: 'react-native'
    });
  }

  trackEvent(eventName, properties = {}) {
    if (!this.initialized) return;
    const platform = Platform.OS === 'ios' ? 'mobile_ios' : 'mobile_android';

    this.send('/api/v1/track', {
      projectId: this.projectId,
      sessionId: this.sessionId,
      userId: this.userId,
      platform: platform,
      eventType: 'custom',
      eventName: eventName,
      properties: properties,
      clientSdk: 'react-native'
    });
  }

  trackError(error, context = '') {
    if (!this.initialized) return;
    const platform = Platform.OS === 'ios' ? 'mobile_ios' : 'mobile_android';

    this.send('/api/v1/track', {
      projectId: this.projectId,
      sessionId: this.sessionId,
      platform: platform,
      eventType: 'error',
      eventName: 'React Native Error',
      path: context,
      error: {
        message: error ? error.message || String(error) : 'Unknown error',
        stack: error ? error.stack || '' : ''
      },
      clientSdk: 'react-native'
    });
  }

  identify(userId) {
    this.userId = userId;
  }

  heartbeat() {
    if (!this.initialized) return;
    this.send('/api/v1/heartbeat', {
      projectId: this.projectId,
      sessionId: this.sessionId
    });
  }

  send(endpoint, data) {
    fetch(`${this.serverUrl}${endpoint}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Platform': Platform.OS === 'ios' ? 'mobile_ios' : 'mobile_android',
        'X-Client-Sdk': 'react-native'
      },
      body: JSON.stringify(data)
    }).catch(() => {});
  }
}

const AIOTracker = new AIOTrackerClient();
export default AIOTracker;
